GPT/gpt-function-calling.py

```diff
@@ -0,0 +1,185 @@
+                            }
+                        },
+                        "required": ["str"],
+                    },
+                },
+            },
+            {
+                "type": "function",
+                "function": {
+                    "name": "search_for_command",
+                    "description": "search_for_command(str: string) - this searches for a command in the VSCode command palette. If I ask you to do something, please use this command to search for an appropriate command.",
+                    "parameters": {
+                        "type": "object",
+                        "properties": {
+                            "str": {
+                                "type": "string",
+                                "description": "The command to search for",
+                            }
+                        },
+                        "required": ["str"],
+                    },
+                },
+            }
+        ],
+        "max_tokens": 2024,
+        "temperature": 0.6,
+        "n": 1,
+        "stop": None,
+        "model": settings.get("user.openai_model"),
+    }
+
+
+    response = requests.post(url, headers=headers, data=json.dumps(data))
+
+    if response.status_code == 200:
+        notify("GPT Task Completed")
+        print(response.json())
+
+        process_function_calls(insert_response, response)
+
+        content = (response.json()["choices"][0]["message"]["content"] or "").strip()
+        if len(content) != 0:
+            actions.user.display_response(content)
+
+    else:
+        notify("GPT Failure: Check API Key, Model, or Prompt")
+        print(response.json())
+
+def process_function_calls(insert_response, response):
+    try:
+        tool_calls = response.json()["choices"][0]["message"]["tool_calls"]
+        while tool_calls:
+            tool = tool_calls.pop()
+            first_argument = tool['function']['arguments']
+            try:
+                first_argument = json.loads(first_argument)['str']
+            except Exception as e:
+                notify(f"Argument Jason was malformed: {e}")
+            match tool['function']['name']:
+                case 'insert':
+                    insert_response(first_argument)
+                case 'display':
+                    actions.user.display_response(first_argument)
+                case 'notify':
+                    actions.user.notify_user(first_argument)
+                case 'search_for_command':
+                    actions.user.search_for_command(first_argument)
+    except Exception as e:
+        notify(f"No tool_calls found in response from LLM: {e}")
+
+@mod.action_class
+class UserActions:
+    def gpt_go(utterance: str, selected_text: str) -> str:
+        """Answer an arbitrary question"""
+
+        return gpt_function_query(utterance, selected_text, actions.user.paste)
+
+    def gpt_go_cursorless(utterance: str, text_to_process: str, cursorless_destination: any):
```

Comment:
you're using `cursorless_get_text_list` so `text_to_process` has type `list[str]`. This makes sense if you want to support multiple targets (eg `"model go <prompt> air and bat and cap after drum"` or `"model go <prompt> every token after bat"`.

If you just want to support single target, you can use `cursorless_get_text` instead

Reply:
so I guess it just gets turned into a string representation of the list I guess. I'd make the type here be `list[str]` and then explicitly convert it to a list here.